packages/endo/src/assemble.js

```diff
@@ -0,0 +1,68 @@
+/* global Compartment */
+
+import { resolve } from "./node-module-specifier.js";
+
+const { entries } = Object;
+
+const defaultCompartment = Compartment;
+
+// q, as in quote, for strings in error messages.
+const q = JSON.stringify;
+
+// Assemble a DAG of compartments as declared in a compartment map starting at
+// the named compartment and building all compartments that it depends upon,
+// recursively threading the modules exported by one compartment into the
+// compartment that imports them.
+// Returns the root of the compartment DAG.
+// Does not load or execute any modules.
+// Uses makeImportHook with the given "root" string of each compartment in the
+// DAG.
+// Passes the given endowments and external modules into the root compartment
+// only.
+export const assemble = ({
+  name,
+  compartments,
+  makeImportHook,
+  parents = [],
+  loaded = {},
+  endowments = {},
+  modules = {},
+  Compartment = defaultCompartment
+}) => {
+  const descriptor = compartments[name];
+  if (descriptor === undefined) {
+    throw new Error(
+      `Cannot assemble compartment graph with missing compartment descriptor named ${q(
+        name
+      )}, needed by ${parents.map(q).join(", ")}`
+    );
+  }
+  const result = loaded[name];
+  if (result !== undefined) {
+    return result;
+  }
+  if (parents.includes(name)) {
+    throw new Error(`Cannot assemble compartment graph that includes a cycle`);
+  }
+
+  for (const [inner, outer] of entries(descriptor.modules || {})) {
+    const { compartment: compartmentName, module: moduleSpecifier } = outer;
+    const compartment = assemble({
+      name: compartmentName,
+      compartments,
+      makeImportHook,
+      parents: [...parents, name],
+      loaded,
+      Compartment
+    });
+    modules[inner] = compartment.module(moduleSpecifier);
```

Comment:
The initial caller can pass `modules` which will be made available to the top-level Compartment, in addition to modules for all of its dependencies (specifically all of the exported modules for all of its dependency packages). Cool. But the dependency modules will shadow any provided by the caller, so this feature couldn't be used for e.g. dependency injection. Is that intentional? I can imagine there's a good reason to do it one way or the other, or maybe even both, but I don't know what those reasons might be.

Reply:
My intent is for this to be usable for dependency injection. I’ll look more closely and will make sure it’s tested.

Reply:
Checked. As written, this is fine for injecting built-in modules that are not declared dependencies, but not for replacing third-party dependencies with overrides.

I think I will land as-is and revisit. I think I want an explicit directive in the compartment map when a package should receive an override. The same mechanism would be used to thread built-in modules expressly to third-party packages and would be driven by policies in package.json or a lavamoat config file.